### PR TITLE
refactor(segment): replace custom implementation of Segment calls by Segment's official SDK

### PR DESCRIPTION
**Pull request by gauthierpetetin**
## **Description**

The purpose of this PR is to replace custom implementation of Segment calls by Segment's official SDK.

## **Changelog**

CHANGELOG entry: null

## **Related issues**

Fixes: https://github.com/MetaMask/MetaMask-planning/issues/7200

## **Manual testing steps**

1. Build the Extension
2. Go through on-boarding
3. Before opting in to MetaMetrics, find your `metametricsId` in MetaMetricsController state (right click > inspect > Application > Extension storage)
4. Confirm you don't see any event in Segment for your `metametricsId`
5. Opt in to MetaMetrics
6. Confirm you see events of type `identify`, `track`, and `page` in Segment for your `metametricsId`

### Edge cases to test

- **Anonymous events**: Click on Send > Continue. You should see both "Transaction Added" and "Transaction Added Anon" events in Segment:
  - "Transaction Added Anon" event should have `anonymousId` set to '0x0000000000000000' and should include sensitive properties (gas_limit, max_fee_per_gas, max_priority_fee_per_gas, etc.).
  - "Transaction Added" should have `userId` set to your `metametricsId` and should not include sensitive properties.
- **Opt out**: Go to Settings > Privacy > Participate in MetaMetrics, opt out from MetaMetrics, and confirm you no longer see events being emitted in Segment.
- **Early event**: Delete the vault (right click > inspect > Application > Extension storage > KeyringController > vault) and restart Extension. You should see "Vault Corruption Detected" event in Sentry.
- **Persistence of non-emitted events**: Opt out from MetaMetrics, do a series of actions that are supposed to trigger events. Confirm these non-emitted events are persisted in state (right click > inspect > Application > Extension storage > MetaMetricsController > segmentApiCalls)

### Segment mock server

If you want to test the Segment mock server, run this command in a console:
```
node development/mock-segment.js
```
Then add the following to .metamaskrc and re-build the extension:
```
SEGMENT_WRITE_KEY=FAKE
SEGMENT_HOST=http://localhost:9090
```

## **Screenshots/Recordings**

NA

## **Pre-merge author checklist**

- [x] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

---

> [!NOTE]
> **Medium Risk**
> Swaps the core analytics transport from a homegrown implementation to `@segment/analytics-node`, plus new build/LavaMoat allowances for `jose`/SDK globals; this could affect event delivery and bundling if configuration is wrong.
> 
> **Overview**
> **Replaces the in-repo Segment implementation** (`analytics.ts` + JS `segment/index.js`) with a TypeScript `segment` module that wraps `@segment/analytics-node` behind a `SegmentClient` interface and keeps `createSegmentMock` for test/no-write-key builds.
> 
> **Adds safeguards and tests around the new shim**, including deep-cloning `track`/`identify`/`page` payloads before forwarding (to avoid failures when controller state is auto-frozen) and a new `segment-shim.test.ts`; `MetaMetricsController` and its tests are updated to depend on `SegmentClient` types.
> 
> **Updates build/security configuration for the new dependency tree**, including transpiling `jose` for browserify and expanding LavaMoat policies/overrides to allow required globals/packages; `package.json`/`yarn.lock` drop the old validation/retry utilities and add `@segment/analytics-node` (plus new transitive deps like `jose`, `dset`, and `@lukeed/uuid`).
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 5efc74d9d9740f9aa5ab4ebc164b9e38f4f15e21. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by gauthierpetetin** [comment]
@SocketSecurity ignore npm/@​segment/​analytics-node@​3.0.0

This package is the official Segment SDK which requires fetch function to call Segment API.

File: package.json [hunk 1]
```diff
@@ -246,7 +246,11 @@
     "tslib@npm:^2.3.0": "~2.6.0",
     "tslib@npm:^2.3.1": "~2.6.0",
     "tslib@npm:^2.4.0": "~2.6.0",
+    "tslib@npm:^2.4.1": "~2.6.0",
     "tslib@npm:^2.6.2": "~2.6.0",
+    "tslib@npm:^2.7.0": "~2.6.0",
+    "tslib@npm:^2.8.0": "~2.6.0",
+    "tslib@npm:^2.8.1": "~2.6.0",
```
Comment (gauthierpetetin) [inline package.json:254]: Note for reviewers: `@segment/analytics-node` pulled in `tslib 2.8.x`, whose compiled `__generator` assigns `Symbol.iterator` on objects inheriting from `Iterator.prototype`, which throws under SES lockdown in browserify LavaMoat builds (webpack only locks down a couple of chunks), so we extended the existing `tslib ~2.6.0` resolutions to cover the new dependency ranges and keep the SDK on the same safe `tslib` as the rest of the repo.

**Comment by gauthierpetetin** [comment]
@SocketSecurity ignore npm/@segment/analytics-node@3.0.0

**Review by itsyoboieltr — DISMISSED** [review]

**Comment by DDDDDanica** [comment]
Tested locally, all worked as expected 👍🏻  @gauthierpetetin do you wanna add a bit test coverage to pass sonar cloud? can give approval when PR is ready 🙏🏻

**Review by DDDDDanica — DISMISSED** [review]

**Review by itsyoboieltr — DISMISSED** [review]

File: lavamoat/webpack/mv3/main/policy.json [hunk 2]
```diff
@@ -1773,15 +1773,44 @@
...
+        "URL": true,
+        "WebSocketPair": true,
+        "WorkerGlobalScope": true,
+        "clearTimeout": true,
+        "console.warn": true,
+        "fetch": true,
+        "importScripts": true,
+        "process": true,
+        "setTimeout": true
+      },
```
Comment (MajorLift) [inline lavamoat/webpack/mv3/main/policy.json:1807]: - `fetch` permission unavoidable as package directly replaces `fetch` call previously made here: https://github.com/MetaMask/metamask-extension/pull/42056/files#diff-9ca9af983f3ca1f9c40374f502eb49484b9600b08e929df6d5c5da4dac5d5cc3L306
- `setTimeout`, `clearTimeout`: likely necessary
- Possibly avoidable? `EdgeRuntime`, `WebSocketPair`, `WorkerGlobalScope`, `importScripts`

File: lavamoat/webpack/mv3/main/policy.json [hunk 3]
```diff
@@ -1773,15 +1773,44 @@
         "@metamask/eth-hd-keyring>ethereum-cryptography>@scure/bip32>@scure/base": true
       }
     },
-    "@segment/loosely-validate-event": {
+    "@segment/analytics-node>@segment/analytics-core": {
       "globals": {
-        "Buffer.byteLength": true
+        "clearTimeout": true,
+        "console": true,
+        "setTimeout": true
```
Comment (MajorLift) [inline lavamoat/webpack/mv3/main/policy.json:1780]: `clearTimeout`, `setTimeout` likely necessary?

File: lavamoat/webpack/mv3/main/policy.json [hunk 4]
```diff
@@ -3365,6 +3394,16 @@
...
+    "@segment/analytics-node>jose": {
+      "globals": {
+        "CryptoKey": true,
+        "TextDecoder": true,
+        "TextEncoder": true,
+        "atob": true,
+        "btoa": true,
+        "crypto": true
+      }
+    },
```
Comment (MajorLift) [inline lavamoat/webpack/mv3/main/policy.json:3406]: crypto globals

**Review by MajorLift — COMMENTED** [review]
@naugtur @MetaMask/policy-reviewers Requesting review on policies for new Segment package.

**Review by MajorLift — COMMENTED** [review]
@naugtur @MetaMask/policy-reviewers Requesting review on policies for new Segment package.

File: lavamoat/webpack/mv3/main/policy.json (hunk 2 above)
Comment (gauthierpetetin) [reply lavamoat/webpack/mv3/main/policy.json:1807]: Fixed in this commit: https://github.com/MetaMask/metamask-extension/pull/42056/changes/cb59738407903aad6ed0869bbcc6ee802ffe011e

File: lavamoat/webpack/mv3/main/policy.json (hunk 4 above)
Comment (gauthierpetetin) [reply lavamoat/webpack/mv3/main/policy.json:3406]: `@segment/analytics-node`’s [Publisher](https://github.com/segmentio/analytics-next/blob/master/packages/node/src/plugins/segmentio/publisher.ts#L95) requires [lib/token-manager](https://github.com/segmentio/analytics-next/blob/master/packages/node/src/plugins/segmentio/publisher.ts#L10), and [token-manager requires jose](https://github.com/segmentio/analytics-next/blob/master/packages/node/src/lib/token-manager.ts#L3) . Even though OAuth is never configured in MetaMask, jose is still bundled and evaluated whenever that Publisher stack loads.
I tried setting these to false, but MetaMask fails to start when I do so, showing an infinite loading spinner and a SES_UNCAUGHT_EXCEPTION in the console.

Image

File: lavamoat/webpack/mv3/main/policy.json (hunk 3 above)
Comment (gauthierpetetin) [reply lavamoat/webpack/mv3/main/policy.json:1780]: `setTimeout` and `clearTimeout` are required by `@segment/analytics-core` for the event queue, priority queue, and callback helpers. Segment analytics in MetaMask relies on it, so the policy grants are appropriate.

Here's the path:
MetaMetricsController.#track
→ MetaMetricsController.#submitSegmentAPICall
→ segment.track (MetaMask Segment shim)
→ @segment/analytics-node Analytics.track
→ Analytics._dispatch
→ dispatchAndEmit
→ @segment/analytics-core dispatch
→ invokeCallback
→ pTimeout: https://github.com/segmentio/analytics-next/blob/master/packages/core/src/callback/index.ts#L6
→ setTimeout, clearTimeout

**Review by MajorLift — APPROVED** [review]
Requested policy changes applied

**Review by DDDDDanica — APPROVED** [review]

**Review by itsyoboieltr — APPROVED** [review]
